# sQTL Test

In this document I explain the code found in [006c\_sQTL\_test.R](006c_sQTL_test.R), where I look at all the potential sQTLs for the inclusion of a particular exon and test whether they are significantly associated with exon inclusion. [006c\_sQTL\_test.R](006c_sQTL_test.R) is used by [006a\_test\_all\_sqtl\_candidates.sh](006a_test_all_sqtl_candidates.sh), which iterates through all the splicing events to be tested.

All the code in this document is written in R.


## 1. Load libraries & files

First, load the libraries that are used in this script:

```r
# library to perform likelihood-ratio test
library(lmtest)

# data table library for data handling
library(data.table)
```
Load the R object generated earlier by [005b\_gtex\_annotations.R](005b_gtex_annotations.R):

```r
# Load PSI.Estimates and GTEx annotations
load("../../Filtered_Annotations_PSIestimates.RData")
```
Load the mini BED file containing information only about this particular exon skipping event (generated by [006a\_test\_all\_sqtl\_candidates.sh](006a_test_all_sqtl_candidates.sh) just before running this R script):

```r
# load bed file with information about this particular splicing event
Splicing.Events <- fread(input = "Gene_Regions.bed")
colnames(Splicing.Events) <- c("Chr", "Start", "End", "ID", "Gene")
```
Load the parsed genotypes file (generated by [006b\_process\_vcf.awk](006b_process_vcf.awk) just before running this R script) that contains (1) different sQTL candidates in different rows and (2) different samples in different columns:

```r
# load parsed genotype file for this particular splicing event
Genotypes <- fread("genotypes.MAF01.GeneRegions.Parsed.txt",
                   skip = "#CHROM")
colnames(Genotypes) <- sapply(X = colnames(Genotypes),
                              FUN = function(x){
                                gsub(pattern = "-", replacement = "\\.", x = x)
                              })

# filter genotypes file to keep only subjects included in our analysis
Subjects.We.Can.Look.At <- which(colnames(Genotypes) %in% GTEX.Dataset.Annotations$Subject.ID)
Genotypes <- Genotypes[,.SD,,.SDcols=c(1,2,3,Subjects.We.Can.Look.At)]
```

## 2. Test for sQTL

Before we perform any tests, we will extract all the PSI values corresponding to the splicing event to be analysed:

```r
# inclusion levels for this particular exon skipping event
This.Splicing.Event <- as.character(Splicing.Events$ID)
This.Event.PSI <- as.numeric(PSI.Estimates[This.Splicing.Event,])
```

The remaining code in the script is embedded within the following `if` statement:

```r
if (nrow(Genotypes) != 0) {
    # code here
}
```
This ensures that we have at least one snp to test. If there were no common variants inside the same gene as the exon of interest, we won't perform any statistical tests. Once we pass this test, we'll create a data frame (`This.Splicing.Event.Table`) for this splicing event containing the information from both `GTEX.Dataset.Annotations` and `This.Event.PSI`:

```r
# build a df for this particular exon splicing event
This.Splicing.Event.Table <- GTEX.Dataset.Annotations
This.Splicing.Event.Table$PSI <- This.Event.PSI
```
Also initialise an empty list which we will fill with the results of our statistical test:

```r
# start an empty list
sQTL.List <- vector(mode = "list", length = nrow(Genotypes))
```
We're now going to loop through the different sQTL candidates (rows in `Genotypes` table) and test whether each of them is significantly associated with exon inclusion:

```r
# loop through all the potential sQTLs and perform a statistical test if appropriate
for (j in 1:nrow(Genotypes)){
    # code here
}
```
We will add a column named 'Genotype' to `This.Splicing.Event.Table`. This column will contain a vector of 0's, 1's and 2's that represent the number of copies of the variant/snp tested in each sample:

```r
# start an empty vector
This.Splicing.Event.Genotypes <- c()

# fill in empty vector with genotypes (0/1/2)
for(each.id in This.Splicing.Event.Table$Subject.ID) {
  if (each.id %in% colnames(Genotypes)) {
    This.Splicing.Event.Genotypes <- c(This.Splicing.Event.Genotypes,
                                       as.numeric(Genotypes[j,
                                                            each.id,
                                                            with=FALSE]))
  } else {
    This.Splicing.Event.Genotypes <- c(This.Splicing.Event.Genotypes,
                                       NA)
  }
}

# put genotypes in new column of This.Splicing.Event.Table
This.Splicing.Event.Table$Genotype <- This.Splicing.Event.Genotypes
```
For this sQTL candidate, I will perform a statistical test separately in all the different tissues for which I have enough samples. Before I do this, however, I need to create an empty list where I'll store the results for this sQTL, sorted by tissue:

```r
# what tissues do I have?
Vector.of.Tissues.Analysed <- unique(as.character(This.Splicing.Event.Table$Tissue))

# start an empty list for p values sorted by tissue
Tissues.List <- vector(mode = "list",
                       length = length(Vector.of.Tissues.Analysed))
                       
# elements of the list named after the different tissues in our dataset
names(Tissues.List) <- Vector.of.Tissues.Analysed
```
Next, with a for loop, we will go tissue by tissue, performing a statistical test if appropriate:

```r
for (each.tissue in Vector.of.Tissues.Analysed) {
    # code here
}
```
For each tissue, we will:

* Create a sub-dataframe containing only the rows of `This.Splicing.Event.Table` corresponding to the relevant tissue.
* Check whether we have enough genotype diversity (at least two genotype categories out of 0, 1 and 2; and at least two categories with > 7 observations).
* Perform a generalised linear model using genotype, ischaemic time, sex and age as predictors.
* Perform a generalised linear model without genotype as a predictor.
* Do a likelihood-ratio test
* Save the p value inside `Tissues.List`

```r
# a sub-dataframe only with rows corresponding to this tissue
SubDF <- This.Splicing.Event.Table[which(as.character(This.Splicing.Event.Table$Tissue) == each.tissue),]
SubDF <- SubDF[complete.cases(SubDF),]

# only do a statistical test if we have at least two of (0, 1, 2);
# and if the no. of observations in at least two of the categories is > 7
if (length(table(SubDF$Genotype)) >= 2 & sum(table(SubDF$Genotype) > 7) >= 2) {
  
  # value to be predicted by the model
  lm.psi <- SubDF$PSI
  lm.psi <- round(lm.psi*100)
  lm.psi <- cbind(lm.psi, 100-lm.psi)
  
  # genotype predictor
  lm.genotype <- as.numeric(as.character(SubDF$Genotype))
  
  # ischaemic time predictor
  # (divide by 1000 because function crashes with large numbers)
  lm.ischaemic <- SubDF$Ischaemic.Time/1000
  
  # sex predictor
  lm.sex <- as.numeric(as.character(SubDF$Sex))
  
  # age predictor
  # (divide by 100 since function crashes with large numbers)
  lm.age <- as.numeric(as.character(SubDF$Age))/100
  
  # model including genotype variable
  res<-try(suppressMessages(glm(lm.psi ~ lm.genotype + lm.ischaemic + lm.sex + lm.age,
                                family=binomial)),
           silent=TRUE)
  
  # model without genotype variable
  res0<-try(suppressMessages(glm(lm.psi ~ lm.ischaemic + lm.sex + lm.age,
                                 family=binomial)),
            silent=TRUE)
  
  # do a likelihood-ratio test and take p value
  P.Value <- lrtest(res,res0)$"Pr(>Chisq)"[2]
  Estimate <- summary(res)$coefficients[2,1]
  
  Tissues.List[[each.tissue]] <- c(P.Value, Estimate)
} else {
  Tissues.List[[each.tissue]] <- NA
}

```
Store `Tissues.List` inside `sQTL.List`:

```r
# put Tissues.List inside sQTL.List
sQTL.List[[j]] <- Tissues.List
```
And finally, after running this code through the various sQTL candidates which might have been tested, we save all the results for later use:

```r
# name elements of sQTL.List with the sQTL IDs
names(sQTL.List) <- Genotypes$ID

# save sQTL.List
save(sQTL.List, file = "sQTL.RData")
```